File: internal/scaffold/compiler_test.go
package scaffold

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func TestMinCompilers(t *testing.T) {
	cases := map[int][2]int{
		23: {15, 17},
		20: {11, 16},
		17: {7, 5},
		14: {5, 4},
		11: {5, 4},
	}
	for std, want := range cases {
		gcc, clang := MinCompilers(std)
		if gcc != want[0] || clang != want[1] {
			t.Errorf("MinCompilers(%d) = (%d, %d), want (%d, %d)", std, gcc, clang, want[0], want[1])
		}
	}
}

func TestCompilerChoices(t *testing.T) {
	// C++23's baseline GCC (15) equals the fetched newest, so it is the sole option.
	gcc, clang := CompilerChoices(23, 15, 20)
	if len(gcc) != 1 || gcc[0] != 15 {
		t.Errorf("CompilerChoices(23) gcc = %v, want [15]", gcc)
	}
	// Clang ranges from its baseline up to the newest, oldest first.
	if len(clang) == 0 || clang[0] != 17 || clang[len(clang)-1] != 20 {
		t.Errorf("CompilerChoices(23) clang = %v, want 17..20", clang)
	}
	// A lower standard opens up more (older) GCC options, baseline first, and the
	// ceiling tracks whatever newest we discovered.
	gcc, _ = CompilerChoices(20, 16, 20)
	if gcc[0] != 11 || gcc[len(gcc)-1] != 16 {
		t.Errorf("CompilerChoices(20, newest 16) gcc = %v, want 11..16", gcc)
	}
	// A newest that lags the baseline (shouldn't happen, but be safe) collapses to
	// the baseline alone rather than yielding an empty list.
	gcc, _ = CompilerChoices(23, 14, 20)
	if len(gcc) != 1 || gcc[0] != 15 {
		t.Errorf("CompilerChoices(23, newest 14) gcc = %v, want [15]", gcc)
	}
}

func TestParseGCCNewest(t *testing.T) {
	index := `<a href="gcc-4.8.5/">gcc-4.8.5/</a>
<a href="gcc-14.2.0/">gcc-14.2.0/</a>
<a href="gcc-15.1.0/">gcc-15.1.0/</a>
<a href="summit/">summit/</a>`
	if got := parseGCCNewest([]byte(index)); got != 15 {
		t.Errorf("parseGCCNewest = %d, want 15", got)
	}
	if got := parseGCCNewest([]byte("no versions here")); got != 0 {
		t.Errorf("parseGCCNewest(none) = %d, want 0", got)
	}
}

func TestParseClangNewest(t *testing.T) {
	body := `[
		{"tag_name":"llvmorg-21.0.0-rc1","prerelease":true},
		{"tag_name":"llvmorg-20.1.8","prerelease":false},
		{"tag_name":"llvmorg-19.1.7","prerelease":false}
	]`
	if got := parseClangNewest([]byte(body)); got != 20 {
		t.Errorf("parseClangNewest = %d, want 20 (skipping the prerelease 21)", got)
	}
	if got := parseClangNewest([]byte("not json")); got != 0 {
		t.Errorf("parseClangNewest(bad) = %d, want 0", got)
	}
}

func TestCompilerGuard(t *testing.T) {
	guard := CompilerGuard(15, 17)
	if !strings.HasPrefix(guard, GuardStart) || !strings.HasSuffix(guard, GuardEnd) {
		t.Fatalf("guard not wrapped in markers:\n%s", guard)
	}
	for _, want := range []string{
		`CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 15`,
		`CMAKE_CXX_COMPILER_ID STREQUAL "Clang" AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 17`,
		"requires GCC >= 15",
		"requires Clang >= 17",
	} {
		if !strings.Contains(guard, want) {
			t.Errorf("guard missing %q\n---\n%s", want, guard)
		}
	}
}

func TestCompilerGuardZeroDisables(t *testing.T) {
	// A zero version drops that compiler's branch; both zero drops the if entirely.
	only := CompilerGuard(15, 0)
	if strings.Contains(only, "Clang") {
		t.Errorf("clang=0 should not emit a Clang branch:\n%s", only)
	}
	if !strings.Contains(only, "GNU") {
		t.Errorf("gcc=15 should emit a GNU branch:\n%s", only)
	}
	none := CompilerGuard(0, 0)
	if strings.Contains(none, "if(") || strings.Contains(none, "FATAL_ERROR") {
		t.Errorf("both zero should emit no check:\n%s", none)
	}
	// Still delimited so `cup compiler` can rewrite it later.
	if !strings.HasPrefix(none, GuardStart) || !strings.HasSuffix(none, GuardEnd) {
		t.Errorf("empty guard must keep its markers:\n%s", none)
	}
}

func TestReplaceCompilerGuard(t *testing.T) {
	root := t.TempDir()
	path := filepath.Join(root, "CMakeLists.txt")
	body := "project(demo)\n\n" + CompilerGuard(15, 17) + "\n\nset(CMAKE_CXX_STANDARD 23)\n"
	if err := os.WriteFile(path, []byte(body), 0o644); err != nil {
		t.Fatal(err)
	}

	if err := ReplaceCompilerGuard(root, path, 13, 0); err != nil {
		t.Fatalf("ReplaceCompilerGuard: %v", err)
	}
	got, _ := os.ReadFile(path)
	out := string(got)
	if !strings.Contains(out, "VERSION_LESS 13") {
		t.Errorf("guard not lowered to 13:\n%s", out)
	}
	if strings.Contains(out, "VERSION_LESS 17") || strings.Contains(out, "Clang") {
		t.Errorf("clang=0 should have dropped the Clang branch:\n%s", out)
	}
	// Surrounding lines are preserved.
	if !strings.Contains(out, "project(demo)") || !strings.Contains(out, "set(CMAKE_CXX_STANDARD 23)") {
		t.Errorf("surrounding CMake lines lost:\n%s", out)
	}
}

func TestReplaceCompilerGuardNoMarkers(t *testing.T) {
	root := t.TempDir()
	path := filepath.Join(root, "CMakeLists.txt")
	if err := os.WriteFile(path, []byte("project(demo)\n"), 0o644); err != nil {
		t.Fatal(err)
	}
	if err := ReplaceCompilerGuard(root, path, 15, 17); err == nil {
		t.Error("ReplaceCompilerGuard on a file without markers = nil error, want error")
	}
}

File: src/tests/compiler_test.cpp
#include <catch2/catch_test_macros.hpp>

#include <filesystem>
#include <fstream>
#include <string>
#include <vector>

#include "TempDir.hpp"

import cup.scaffold;

namespace {

using cup::test::TempDir;

[[nodiscard]] std::string read(const std::string& path) {
    std::ifstream in(path, std::ios::binary);
    return std::string(std::istreambuf_iterator<char>(in), std::istreambuf_iterator<char>());
}

}

TEST_CASE("min_compilers", "[compiler][min_compilers]") {
    struct Case {
        int std;
        int gcc;
        int clang;
    };
    for (const auto& c : {Case{23, 15, 17}, Case{20, 11, 16}, Case{17, 7, 5}, Case{14, 5, 4},
                          Case{11, 5, 4}}) {
        INFO("std: " << c.std);
        const auto [gcc, clang] = cup::scaffold::min_compilers(c.std);
        REQUIRE(gcc == c.gcc);
        REQUIRE(clang == c.clang);
    }
}

TEST_CASE("compiler_choices", "[compiler][compiler_choices]") {
    // C++23's baseline GCC (15) equals the fetched newest, so it is the sole option.
    {
        const auto [gcc, clang] = cup::scaffold::compiler_choices(23, 15, 20);
        REQUIRE(gcc == std::vector<int>{15});
        // Clang ranges from its baseline up to the newest, oldest first.
        REQUIRE_FALSE(clang.empty());
        REQUIRE(clang.front() == 17);
        REQUIRE(clang.back() == 20);
    }
    // A lower standard opens up more (older) GCC options, baseline first, and the
    // ceiling tracks whatever newest we discovered.
    {
        const auto [gcc, clang] = cup::scaffold::compiler_choices(20, 16, 20);
        (void)clang;
        REQUIRE(gcc.front() == 11);
        REQUIRE(gcc.back() == 16);
    }
    // A newest that lags the baseline (shouldn't happen, but be safe) collapses to
    // the baseline alone rather than yielding an empty list.
    {
        const auto [gcc, clang] = cup::scaffold::compiler_choices(23, 14, 20);
        (void)clang;
        REQUIRE(gcc == std::vector<int>{15});
    }
}

TEST_CASE("compiler_guard", "[compiler][compiler_guard]") {
    const std::string guard = cup::scaffold::compiler_guard(15, 17);
    REQUIRE(guard.starts_with(cup::scaffold::kGuardStart));
    REQUIRE(guard.ends_with(cup::scaffold::kGuardEnd));
    for (const auto* want : {
             R"(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 15)",
             R"(CMAKE_CXX_COMPILER_ID STREQUAL "Clang" AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 17)",
             "requires GCC >= 15",
             "requires Clang >= 17",
         }) {
        INFO("expected substring: " << want);
        REQUIRE(guard.find(want) != std::string::npos);
    }
}

TEST_CASE("compiler_guard: a zero version disables that compiler's branch",
          "[compiler][compiler_guard]") {
    // A zero version drops that compiler's branch; both zero drops the if entirely.
    const std::string only = cup::scaffold::compiler_guard(15, 0);
    REQUIRE(only.find("Clang") == std::string::npos);
    REQUIRE(only.find("GNU") != std::string::npos);

    const std::string none = cup::scaffold::compiler_guard(0, 0);
    REQUIRE(none.find("if(") == std::string::npos);
    REQUIRE(none.find("FATAL_ERROR") == std::string::npos);
    // Still delimited so `cup compiler` can rewrite it later.
    REQUIRE(none.starts_with(cup::scaffold::kGuardStart));
    REQUIRE(none.ends_with(cup::scaffold::kGuardEnd));
}

TEST_CASE("replace_compiler_guard", "[compiler][replace_compiler_guard]") {
    const TempDir root;
    const std::string path = (root.path() / "CMakeLists.txt").string();
    const std::string body = "project(demo)\n\n" + cup::scaffold::compiler_guard(15, 17) +
                             "\n\nset(CMAKE_CXX_STANDARD 23)\n";
    root.write("CMakeLists.txt", body);

    REQUIRE(cup::scaffold::replace_compiler_guard(root.path().string(), path, 13, 0).has_value());
    const std::string out = read(path);
    REQUIRE(out.find("VERSION_LESS 13") != std::string::npos);
    REQUIRE(out.find("VERSION_LESS 17") == std::string::npos);
    REQUIRE(out.find("Clang") == std::string::npos);
    // Surrounding lines are preserved.
    REQUIRE(out.find("project(demo)") != std::string::npos);
    REQUIRE(out.find("set(CMAKE_CXX_STANDARD 23)") != std::string::npos);
}

TEST_CASE("replace_compiler_guard on a file without markers",
          "[compiler][replace_compiler_guard]") {
    const TempDir root;
    const std::string path = (root.path() / "CMakeLists.txt").string();
    root.write("CMakeLists.txt", "project(demo)\n");

    REQUIRE_FALSE(
        cup::scaffold::replace_compiler_guard(root.path().string(), path, 15, 17).has_value());
}
